src/Cooked/MockChain/GenerateTx/Proposal.hs

```diff
@@ -0,0 +1,147 @@
+        GovActionLifetime n -> setL Conway.ppuGovActionLifetimeL $ Cardano.EpochInterval $ fromIntegral n
+        GovActionDeposit n -> setL Conway.ppuGovActionDepositL $ fromIntegral n
+        DRepRegistrationDeposit n -> setL Conway.ppuDRepDepositL $ fromIntegral n
+        DRepActivity n -> setL Conway.ppuDRepActivityL $ Cardano.EpochInterval $ fromIntegral n
+
+-- | Translates a given skeleton proposal into a governance action
+toGovAction :: TxSkelProposal -> ProposalGen (Conway.GovAction Emulator.EmulatorEra)
+toGovAction TxSkelProposal {..} = do
+  sHash <- case txSkelProposalWitness of
+    Nothing -> return SNothing
+    Just (script, _) -> do
+      Cardano.ScriptHash sHash <-
+        throwOnToCardanoError
+          "Unable to convert script hash"
+          (Ledger.toCardanoScriptHash (toScriptHash script))
+      return $ SJust sHash
+  case txSkelProposalAction of
+    TxGovActionParameterChange changes ->
+      return $
+        Conway.ParameterChange
+          SNothing -- TODO, should not be Nothing later on
+          (foldl (flip toPParamsUpdate) (Conway.PParamsUpdate Cardano.emptyPParamsStrictMaybe) changes)
+          sHash
+    TxGovActionHardForkInitiation _ -> throwOnString "TxGovActionHardForkInitiation unsupported"
+    TxGovActionTreasuryWithdrawals mapCredentialLovelace -> do
+      cardanoMap <- SMap.fromList <$> mapM (\(cred, Api.Lovelace lv) -> (,Emulator.Coin lv) <$> liftTxGen (toRewardAccount cred)) (Map.toList mapCredentialLovelace)
+      return $ Conway.TreasuryWithdrawals cardanoMap sHash
+    TxGovActionNoConfidence -> return $ Conway.NoConfidence SNothing -- TODO, should not be Nothing later on
+    TxGovActionUpdateCommittee {} -> throwOnString "TxGovActionUpdateCommittee unsupported"
+    TxGovActionNewConstitution _ -> throwOnString "TxGovActionNewConstitution unsupported"
+
+-- | Translates a skeleton proposal into a proposal procedure
+-- alongside a possible witness
+toProposalProcedureAndWitness :: TxSkelProposal -> ProposalGen (Conway.ProposalProcedure Emulator.EmulatorEra, Maybe (Cardano.ScriptWitness Cardano.WitCtxStake Cardano.ConwayEra))
+toProposalProcedureAndWitness txSkelProposal@TxSkelProposal {..} = do
+  minDeposit <- asks (Emulator.unCoin . Lens.view Conway.ppGovActionDepositL . fst)
+  cred <- liftTxGen $ toRewardAccount $ toCredential txSkelProposalAddress
+  govAction <- toGovAction txSkelProposal
+  let proposalAnchor = do
+        anchor <- txSkelProposalAnchor
+        anchorUrl <- Cardano.textToUrl (length anchor) (Text.pack anchor)
+        let anchorDataHash =
+              handle
+                (return . throwOnString . (("Error when parsing anchor " ++ show anchor ++ " with error: ") ++) . (show @Network.HttpException))
+                ((Network.parseRequest anchor >>= Network.httpBS) <&> return . Cardano.hashAnchorData . Cardano.AnchorData . Network.getResponseBody)
+        return $ Cardano.Anchor anchorUrl <$> unsafePerformIO anchorDataHash
```

Comment:
So there is actual network access required for this? Should we document this more prominently in cooked? Until now cooked was purely offline.

Reply:
I would actually be in favor of a pure solution. This would break reproducibility. We could instead pre-resolve urls and provide a table associating urls to their content (or Nothing).

This could be implemented in a future PR and the hash verification dropped for now.